data/transactions/logic/eval.go

```diff
@@ -4060,3 +4062,126 @@ func opBase64Decode(cx *EvalContext) {
 	}
 	cx.stack[last].Bytes, cx.err = base64Decode(cx.stack[last].Bytes, encoding)
 }
+func hasDuplicateKeys(dec *json.Decoder) (bool, error) {
+	keys := make(map[string]bool)
+	var value interface{}
+	dec.Token() // {
+	for dec.More() {
+		// Get JSON key
+		key, err := dec.Token()
+		if err != nil {
+			return false, err
+		}
+		// end of json
+		if key == '}' {
+			break
+		}
+		// decode value
+		err = dec.Decode(&value)
+		if err != nil {
+			return false, err
+		}
+		// check for duplicates
+		if _, ok := keys[key.(string)]; ok {
+			return true, nil
+		}
+		keys[key.(string)] = true
+	}
+	return false, nil
+}
+
+func parseJSON(jsonText []byte) (interface{}, error) {
+	if !json.Valid(jsonText) {
+		return nil, fmt.Errorf("invalid json text")
+	}
+	// check for duplicate keys
+	decoder := json.NewDecoder(bytes.NewReader(jsonText))
+	hasDuplicates, err := hasDuplicateKeys(decoder)
+	if hasDuplicates {
+		return nil, fmt.Errorf("invalid json text, duplicate keys not allowed")
+	}
+	// decode text
+	decoder = json.NewDecoder(bytes.NewReader(jsonText))
+	decoder.UseNumber()
+	var parsed interface{}
+	err = decoder.Decode(&parsed)
+	if err != nil {
+		return nil, err
+	}
+	return parsed, nil
+}
+func opJSONRef(cx *EvalContext) {
+	// get json key
+	last := len(cx.stack) - 1
+	key := string(cx.stack[last].Bytes)
+	cx.stack = cx.stack[:last] // pop
+
+	//parse json text
+	last = len(cx.stack) - 1
+	resp, err := parseJSON(cx.stack[last].Bytes)
+	if err != nil {
+		cx.err = fmt.Errorf("error while parsing JSON text, %v", err)
+		return
+	}
+
+	//get value from json
+	var stval stackValue
+	resp, ok := resp.(map[string]interface{})[key]
+	if !ok {
+		cx.err = fmt.Errorf("key %s not found in JSON text", key)
+		return
+	}
+	expectedType := jsonRefTypeNames[cx.program[cx.pc+1]]
+	switch expectedType {
+	case "JSONString":
+		val, ok := resp.(string)
+		if !ok {
+			cx.err = fmt.Errorf("got type %T but expected string", resp)
+			return
+		}
+		stval.Bytes = []byte(val)
+	case "JSONUint64":
+		val, ok := resp.(json.Number)
+		if !ok {
+			cx.err = fmt.Errorf("got type %T but expected uint64", resp)
+			return
+		}
+		// val is not a float
+		if strings.Contains(val.String(), ".") {
+			cx.err = fmt.Errorf("got type float64 but expected uint64")
+			return
+		}
+		// val is not negative
+		if val.String()[0] == '-' {
+			cx.err = fmt.Errorf("JSON value should be a uint64")
+			return
+		}
+		// val > max uint64
+		maxVal := strconv.FormatUint(math.MaxUint64, 10)
+		if val.String() > maxVal {
+			cx.err = fmt.Errorf("JSON value range within uint64 range")
+			return
+		}
+		// convert to uint64
+		var bigInt big.Int
+		bigInt.UnmarshalText([]byte(val.String()))
+		stval.Uint = bigInt.Uint64()
+	case "JSONObject":
+		_, ok := resp.(map[string]interface{})
+		if !ok {
+			cx.err = fmt.Errorf("got type %T but expected JSON object", resp)
+			return
+		}
+		var rawJSONText map[string]json.RawMessage
+		err = json.Unmarshal(cx.stack[last].Bytes, &rawJSONText)
```

Comment:
Instead of unmarshalling a JSON object again, I suggest you make the following changes to your approach:

1. Since this opcode only takes JSON objects as input (not arrays or primitive values), in your `parseJSON` function, don't decode to an `interface{}`, instead decode to a `map[string]json.RawMessage` (and now there's no need to call `decoder.UseNumber()`)

2. Then call `hasDuplicateKeys`, since at this point you are certain that the input is a JSON object.

3. For `JSONString` and `JSONUint64`, get the appropriate `json.RawMessage` value from the decoded map, then do this:
```go
var value string // change to uint64 for JSONUint64
err := json.Unmarshal(json_raw_message_casted_to_bytes_goes_here, &value)
```
I _think_ when you `json.Unmarshal` to a uint64, it uses `strconv.ParseUint` under the hood, so you should get the same result with less code. If not, or if you want to be more explicit, you could decode to a `json.Number` instead.

4. For `JSONObject`, no need to unmarshal anything else, just return the `json.RawMessage` value directly.

As an added bonus, this should be more efficient, since the original object is a `map[string]json.RawMessage`, meaning the other keys that aren't being referenced won't have their values parsed into in memory objects.